dts/bindings/misc/zephyr,flash-disk.yaml

```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Block storage device on flash partition
+
+compatible: "zephyr,flash-disk"
+
+include: base.yaml
+
+properties:
+    partition:
+      type: phandle
+      required: true
+      description: |
+        Backing storage flash map partition.
+
+    disk-name:
+      type: string
+      required: true
+      description: |
+        Disk name.
+
+    sector-size:
+      type: int
+      required: false
+      default: 512
+      description: |
+        Emulated block device sector size in bytes.
+
+    buffer-size:
+      type: int
+      required: true
```

Comment:
Make this optional: it can be obtained from flashcontroller for SoC devices, so this would be override.
Still don't like the name.
In case of SPI devices this also could be made optional, unfortunately the hardcoded sector size is only available through definitions in [zephyr/drivers/flash/spi_nor.h](https://github.com/zephyrproject-rtos/zephyr/blob/c022dd7756209b9ea20228757829ea13b362f7ee/drivers/flash/spi_nor.h#L82)

Reply:
SPI (or rather QSPI) devices are the main target in my opinion. Using `SPI_NOR_SECTOR_SIZE` doesn't sound like a good solution to a problem. The problem here is that partition devicetree entry does not provide erase block size.

Reply:
> SPI (or rather QSPI) devices are the main target in my opinion. Using `SPI_NOR_SECTOR_SIZE` doesn't sound like a good solution to a problem. The problem here is that partition devicetree entry does not provide erase block size.

I am not insisting on using `SPI_NOR_SECTOR_SIZE`, I am just making note here.
The flash subsystem uses it as a compromise. You can use different SPI commands to use other sizes, or query a flash device for supported erase sizes at runtime - but this does not help us with setting up the rmw buffer at compile time.

Reply:
The lack of something that would help setting the buffer size at compile time is the reason why the `zephyr,flash-disk` compatible needs erase size property. I will probably change `buffer-size` to `erase-size` as it is really a property of underlying hardware.

Reply:
I think erase-size is a bad name. Whatever it is used for the main result of the buffer is that it takes space, the user should therefore be aware of this. This is very clear from the name buffer-size, it is not clear at all when it is hidden or when it is given a name like erase-size.

Reply:
Yes, it takes space. But the fact that it takes space is a consequence not the root cause. The root cause is the fact that `erase-size` is the smallest number of bytes that can be erased at a time (and thus to not lose the data, the data has to be stored somewhere). I can add the RAM usage consequence information to description.

Reply:
I agree that the root cause is the erase-size. But think ahead, changes to `flash_disk` might result in the buffer being used for caching or a block needed for flash wear distribution, ... the name `erase-size` might become meaningless and would need to be changed, requiring the change of all dts's that are using this.
As an anology think about the clock line of spi devices, they could be called `data-input-sync-signal` for some devices but naming them like would only make sense for some devices so it is better to just call it `CLK`.

Reply:
I plan to use this exact buffer for caching, and it is really the perfect fit for caching scenario. Basically you need `erase-size` bytes buffer to cache one "erase block". However, the caching itself is more about *software* configuration not hardware limitations/descriptions. Having separate "single block buffer" cache per-disk should really be a Kconfig option. Caching just one block is very low-cost, meaningful-gain solution because the buffer is already there and atleast the USB Mass Storage hosts tend to send multiple sectors at a time.

While wear leveling is important, I think flashdisk is really poor choice - if wear leveling is important, user should really use something different in my opinion. For wear leveling there would be more parameters as it is pretty much impossible to implement it without extra space (flashdisk volume size is currenlty equal to storage partition size; for wear leveling it would have to be less).

Reply:
> I plan to use this exact buffer for caching, and it is really the perfect fit for caching scenario. Basically you need `erase-size` bytes buffer to cache one "erase block". However, the caching itself is more about _software_ configuration not hardware limitations/descriptions. Having separate "single block buffer" cache per-disk should really be a Kconfig option. Caching just one block is very low-cost, meaningful-gain solution because the buffer is already there and atleast the USB Mass Storage hosts tend to send multiple sectors at a time.
>
And what if the caching would be implemented with a double sized buffer using a ping-pong approach to allow filling one part of the cache while using the other part to do the flash writing ? How do you match the buffer size with erase-size (I know it is just times 2), but what is the meaning of `erase-size`. Regarding the Kconfig option, as long as it doesn't double the things that need to be defined in dts anyhow I don't mind using Kconfig, but it needs to be a per instance option. Defining it in the dts is for now the easiest and most elegant solution. As far as I know nobody has volunteered to try and move the partition definitions into Kconfig, they are also a "software" construct.
  
> While wear leveling is important, I think flashdisk is really poor choice - if wear leveling is important, user should really use something different in my opinion. For wear leveling there would be more parameters as it is pretty much impossible to implement it without extra space (flashdisk volume size is currenlty equal to storage partition size; for wear leveling it would have to be less).

While it it's current implementation it is not a good choice, but this has nothing to do with the dts description. The dts describes how much flash is available for use by flash_disk. How it is used is the flash_disk implementation aspect, the disk size (number of sectors) reported will just be smaller in case wear-leveling is used.

Reply:
`buffer-size` was renamed to `erase-size` and now it was renamed to `cache-size`. However, using it cache more than 1 eraseable block, even if it is large enough, is problematic because the cache metadata needs to be stored somewhere.

Reply:
I'm not in favor of exposing `cache-size` as a devicetree property.  Can we not just determine what it needs to be inside the driver itself.

Reply:
@galak, sadly this is not possible. There are soc NOR devices where the block-size is non-uniform and the needed size can only be retrieved at run-time, also for SPI NOR devices the block-size is retrieved at run-time and there are also SPI NOR devices with non-uniform block-sizes. Sizing it the largest block-size of a device is also not a solution because this size will be to large. Allocating a buffer at runtime is possible but this might not be considered a valid solution for embedded applications.

Reply:
Ok, can we expand the description than to explain how one should figure out how to set this value.  (and that we expect a runtime check to ensure its set properly)

Reply:
@galak What do you mean? How is "Size of statically allocated cache buffer size in bytes. Buffer must be large enough to hold complete backing storage erase page worth of data." not enough? How would you phrase it?

The "backing storage" refers to partition ("Backing storage flash map partition.") and "erase page" is what the person writing devicetree is expected to be able to find in datasheet.

There is runtime check that simply fails flashdisk at initialization if the buffer is too small.

Reply:
I think adding something like what @Laczen commented on about "some flashes may support multiple different page sizes and this value should be the largest value.  It is expected that a driver will verify at runtime that the value is large enough compared to the page sizes supported by the flash device."

Reply:
I think this suggestion makes sense. It will help the user understand what the number is actually about.

Reply:
Except currently the flashdisk driver code will fail when the backing storage partition has non-uniform page sizes. I can add a runtime check to verify that the page sizes are uniform and fail if they are not. Currently there is assumption that all the pages are uniform within partition.

Reply:
I am not sure if the binding should even mention the uniform page size limitation. It is flashdisk driver limitation, and therefore I think just runtime error is appropriate.

Note that adding support for non-uniform page sizes within flashdisk storage partition is way out of scope in my opinion (it started out as a simple fix to issue that now has been reported #51720 independently). If you want to push it into the scope of this PR, then well, this probably won't ever be merged as there'll always be something new...

Reply:
@Laczen @galak Could you please check if the latest commit that improves the runtime error check sufficiently relieves your concerns? If not, and you think the binding description itself should be modified, please write the exact wording I should use in the description.

Reply:
@desowin, regarding the description:
`
Size of statically allocated buffer size in bytes. The size should be adequately chosen. On storage backends with uniform erase-blocks it should be at least the erase-block-size, on storage backends with non-uniform erase-blocks it should be at least the largest erase-block-size.
`